### PR TITLE
Secondary Indexes

### PR DESCRIPTION
**Pull request by Hydrocharged**
This has been cleaned up and whatnot, so you can give this an earnest look. Most things are a deletion, since I chose to go about things a bit differently than it was already coded in, but overall there are relatively few new additions. I also removed a lot of the temporary code that was hanging around, as it turns out most of it wasn't needed.

### CODE REVIEW

File: sql/parse/parse.go [hunk 1]
```diff
@@ -492,25 +489,79 @@ func convertRenameTable(ctx *sql.Context, ddl *sqlparser.DDL) (sql.Node, error)
...
+				constraint = sql.IndexConstraint_Fulltext
+			case sqlparser.SpatialStr:
+				constraint = sql.IndexConstraint_Spatial
+			default:
+				constraint = sql.IndexConstraint_None
+			}
+
+			columns := make([]sql.IndexColumn, len(ddl.IndexSpec.Columns))
+			for i, col := range ddl.IndexSpec.Columns {
+				if col.Length != nil {
```
Comment (zachmu) [inline sql/parse/parse.go:520]: Given that length is unsupported, seems like you should remove it from sql.IndexColumn

File: sql/parse/parse.go [hunk 2]
```diff
@@ -492,25 +489,79 @@ func convertRenameTable(ctx *sql.Context, ddl *sqlparser.DDL) (sql.Node, error)
...
-	case sqlparser.AddStr:
-		sch, err := tableSpecToSchema(ctx, ddl.TableSpec)
-		if err != nil {
-			return nil, err
+	if ddl.IndexSpec != nil {
+		table := plan.NewUnresolvedTable(ddl.Table.Name.String(), ddl.Table.Qualifier.String())
+		switch ddl.IndexSpec.Action {
+		case sqlparser.CreateStr:
+			var using sql.IndexUsing
+			switch ddl.IndexSpec.Using.Lowered() {
```
[the remark this answers is not in the document]
Comment (Hydrocharged) [reply sql/parse/parse.go:497]: `Using` is a `ColIdent` struct and not a pointer, so if it is unassigned it'll default to the zero value, which is an empty string in this case.

File: sql/core.go [hunk 3]
```diff
@@ -200,6 +200,29 @@ type ProjectedTable interface {
...
+	IndexConstraint_None IndexConstraint = iota
+	IndexConstraint_Unique
+	IndexConstraint_Fulltext
+	IndexConstraint_Spatial
+)
+
+// IndexColumn is the column by which to add to an index.
+type IndexColumn struct {
+	Name string
+	Length int64
```
[the remark this answers is not in the document]
Comment (Hydrocharged) [reply sql/core.go:223]: When specifying an index, you can append a length to string types, which will then only index a substring of that type.

File: sql/parse/parse.go (hunk 1 above)
Comment (Hydrocharged) [reply sql/parse/parse.go:520]: Honestly that's more of a dolt error. If anything I should probably move the error into dolt, but `go-mysql-server` should pass around key lengths IMO.

File: sql/parse/parse.go [hunk 4]
```diff
@@ -492,25 +491,105 @@ func convertRenameTable(ctx *sql.Context, ddl *sqlparser.DDL) (sql.Node, error)
 }
 
 func convertAlterTable(ctx *sql.Context, ddl *sqlparser.DDL) (sql.Node, error) {
-	switch ddl.ColumnAction {
-	case sqlparser.AddStr:
-		sch, err := tableSpecToSchema(ctx, ddl.TableSpec)
-		if err != nil {
-			return nil, err
+	if ddl.IndexSpec != nil {
```
Comment (zachmu) [inline sql/parse/parse.go:494]: Plz extract methods for the two major code paths here

File: sql/plan/alter_index.go [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+		return NewAlterDropIndex(children[0], p.IndexName), nil
+	case IndexAction_Rename:
+		return NewAlterRenameIndex(children[0], p.PreviousIndexName, p.IndexName), nil
+	default:
+		return nil, ErrIndexActionNotImplemented.New(p.Action)
+	}
+}
+
+func (p AlterIndex) String() string {
+	//TODO: fix this
```
Comment (zachmu) [inline sql/plan/alter_index.go:169]: fix before committing plz

**Pull request merged**